libraries/base/GHC/List.hs

```diff
@@ -947,12 +947,19 @@ foldr2_left  k _z  x  r (y:ys) = k x y (r ys)
 
 ----------------------------------------------
 -- | 'zip' takes two lists and returns a list of corresponding pairs.
+--
+-- > zip [1, 2] ['a', 'b'] = [(1, 'a'), (2, 'b')]
+--
 -- If one input list is short, excess elements of the longer list are
--- discarded.
+-- discarded:
+--
+-- > zip [1] ['a', 'b'] = [(1, 'a')]
+-- > zip [1, 2] ['a'] = [(1, 'a')]
 --
 -- 'zip' is right-lazy:
 --
--- > zip [] _|_ = []
+-- > zip [] undefined = []
```

Comment:
As a beginner I was extremely confused by the difference between `undefined`, `error "..."`, "bottom", `_|_`, and `⊥`. It took me a while to learn that they are all more or less equivalent. I'm not alone either: https://stackoverflow.com/questions/32505911/what-is-the-role-of-bottom-⊥-in-haskell-function-definitions

I changed it here to try and avoid that confusion, and also because this can be put straight into GHCi.

``` hs
>>> zip [] undefined
[]
>>> zip [] _|_
<interactive>:2:11: error:
    parse error (possibly incorrect indentation or mismatched brackets)
```

In the `Prelude` documentation, the symbol `_|_` is referenced by `zip`, `zipWith`, and `splitAt`. It's not defined by any of those things. The term "bottom" is referenced by `seq` but it's not given a precise definition. Neither `undefined` nor `error` mention their relation to `_|_` or "bottom". 

I'm trying to make things a little more concrete, but I'm not going to die on this hill right now. I'd be willing to change `undefined` back to `_|_` if that's desired.

Reply:
That's an entirely different can of worms! It's clear (to me, right now, at least) that the example is meant as an equivalence rather than a Haskell declaration. Even so, `zip [] _|_ = []` is almost valid Haskell. GHCi will give you a helpful hint if you try it out:

``` hs
```>>> zip [] _|_ = []
<interactive>:1:10: error:
    • Found hole: _ :: Bool
    • In the expression: _
      In a stmt of a pattern guard for an equation for ‘zip’: _
      In an equation for ‘zip’: zip [] _ | _ = []
```

I would prefer setting up more typical doctest-style examples, but that seemed too far out of scope. If we wanted to do that, I'd rewrite the examples like this:

``` hs
>>> zip [] undefined
[]

>>> zip undefined []
*** Exception: Prelude.undefined
...
```

Reply:
Ah, whoops, I got confused by the syntax for a moment. I still prefer `_|_` because `undefined` doesn't capture non-termination. (At least, when interpreting it as the Haskell value called `undefined`, which I think is your point).

Reply:
Should I replace `undefined` with `_|_`? I still believe that `undefined` would be better to have in the docs here, but right now I'm more interested in adding the concrete examples of `zip` usage. I just want to get this PR merged quickly.

Reply:
I think if you want to get it merged quickly, it would be smart to leave `_|_` as-is (since as you note, this is a convention seen elsewhere in `base`). However, I applaud the effort to simplify things for beginners, and also thank you for the doc fix as well!

Reply:
I went back to using `_|_` instead of `undefined`. I still think it would be nice for the `Prelude` documentation to explain the relationship between `undefined`, `error "..."`, "bottom", `_|_`, ⊥, and `seq`, but that's for another day 🙂